### PR TITLE
feat: use ETag to avoid configuration loading when it has not changed

### PR DESCRIPTION
**Pull request by typotter**
fixes FF-2726

### Motivation and Context
We don't want to pull flag configuration if it hasn't changed from what we already have parsed and stored. To accomplish this, the server sets the ETag header which changes each time the config changes. The SDK sets the If-None-Match header to whatever ETag it last received to allow the API server to send an empty response body if the config has not changed

### Description

- Create a structured return from the `APIRequestWrapper` that includes the ETag and whether or not the resource `isModified`.
- Do not bother parsing if the resource has not changed (but do set the lastFetch timestamp).
- Store the resource metadata (timestamp and ETag) in the metadata cache (instead of just the last fetch timestamp)

### How is this tested
Unit Tests

### CODE REVIEW

File: src/API/APIRequestWrapper.php [hunk 1]
```diff
@@ -1,6 +1,6 @@
 <?php
 
-namespace Eppo;
+namespace Eppo\API;
```
Comment (typotter) [inline src/API/APIRequestWrapper.php:3]: Moved to API package

File: src/Bandits/BanditVariationIndexer.php [hunk 2]
```diff
@@ -26,11 +26,16 @@ public function __unserialize(array $data): void
      */
     private array $banditFlags = [];
 
+
+    private function __construct()
```
Comment (typotter) [inline src/Bandits/BanditVariationIndexer.php:30]: switching to static helper methods for construction as I wanted to be able to create an empty indexer without having to worry about it throwing an exception

File: src/Config/IConfigurationStore.php [hunk 3]
```diff
@@ -12,21 +12,22 @@ interface IConfigurationStore extends IFlags
...
     /**
-     * Gets the age of the cache
+     * Sets metadata in the data store.
      *
-     * @return int The age of the cache in seconds. -1 if there has been no cache set.
+     * @param string $key
+     * @param mixed $metadata
+     * @return void
      */
-    public function getFlagCacheAgeSeconds(): int;
```
Comment (typotter) [inline src/Config/IConfigurationStore.php:31]: Moved metadata particulars up one level to `ConfigurationLoader`

File: .github/workflows/run-tests.yml [hunk 4]
```diff
@@ -4,7 +4,7 @@ on:
   push:
     branches: [ "main" ]
   pull_request:
-    branches: [ "main" ]
+    branches: [ "*" ]
```
Comment (aarsilv) [inline .github/workflows/run-tests.yml:7]: 👍

File: src/Config/ConfigurationStore.php [hunk 5]
```diff
@@ -79,35 +79,57 @@ private function setFlags(array $flags): void
...
+    /**
+     * @throws InvalidArgumentException
+     */
+    public function setMetadata(string $key, mixed $metadata): void
+    {
+        Validator::validateNotEqual(
+            $key,
+            self::BANDIT_VARIATION_KEY,
+            'Unable to use reserved key, ' . self::BANDIT_VARIATION_KEY
+        );
```
Comment (aarsilv) [inline src/Config/ConfigurationStore.php:128]: Is this just being defensive? We control these keys right--a user won't be able to pick key for meta data?

**Review by aarsilv — APPROVED** [review]

File: src/Config/ConfigurationStore.php (hunk 5 above)
Comment (typotter) [reply src/Config/ConfigurationStore.php:129]: yep, you're right. All these layers got me thinking overly encapsulated.

**Pull request merged**